### ISSUE TITLE
Non-default ABIs aren't supported

### ISSUE THREAD

**Issue by sadlerap** [rust-lang/rustc_codegen_gcc#451]
[The following function miscompiles][godbolt]:

```rust
#[no_mangle]
extern "efiapi" fn efi(x: u32, y: u32) -> u32 {
    x * y
}
```

LLVM emits this:

```asm
efi:
        mov     eax, ecx
        imul    eax, edx
        ret
```

GCC emits this (notice the incorrect registers used for parameters):

```asm
efi:
        mov     eax, edi
        imul    eax, esi
        ret
```

As far as I can tell, we don't preserve the declared ABI since libgccjit doesn't have a way to annotate the ABI of functions with an annotated ABI.  This needs to be implemented for `ui/asm/x86_64/multiple-clobber-abi.rs` to pass, which requires both the `sysv64` and `win64` extern abi to be supported.

[godbolt]: https://godbolt.org/z/vsvYosnbz

**Comment by antoyo** [comment]
Yeah, for that, we would need to support a few new [function attributes](https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html). If you want to take a look, you would need to add the attributes in a few places:

 * https://github.com/antoyo/gcc/blob/8e0113788aca4824e39acbc85237ab275075835d/gcc/jit/libgccjit.h#L2111
 * https://github.com/antoyo/gcc/blob/8e0113788aca4824e39acbc85237ab275075835d/gcc/jit/jit-recording.cc#L4822
 * https://github.com/antoyo/gcc/blob/8e0113788aca4824e39acbc85237ab275075835d/gcc/jit/jit-playback.cc#L523

**Comment by antoyo** [comment]
I believe we [now support most ABIs](https://github.com/rust-lang/rustc_codegen_gcc/blob/5c50a42bab47dd2ff206cc6e5a9b53909004ff5a/src/abi.rs#L241) and the Godbolt example now compiles properly.
